### ISSUE TITLE
[NF/NB] structural array parameters are not getting evaluated

### ISSUE THREAD

**Issue by kabdelhak** [OpenModelica/OpenModelica#12605]
### Description

An element of an array parameter is used as dimension size for a variable and is not getting evaluated. This is needed for the new backend to work.

### Steps to Reproduce

Run model `Modelica.Utilities.Examples.ReadRealMatrixFromFile` from the MSL version `4.0.0` using `--newBackend` flag. It produces following error in the very first phase of the new backend:
```
Error: Internal error NBackendDAE.lowerVariable failed for final parameter Real[3, dim3[2]] A3 = Modelica.Utilities.Streams.readRealMatrix(file3, matrixName1, dim3[1], dim3[2], true)
```

### Expected Behavior

I expect the parameter `dim3[2]` to be evaluated here. From my point of view there are two possible solutions of which i prefer the first one:
- keep the parameter and just replace each structural occurrence with the value
- evaluate the whole of `dim3` even if only one element is structural

It seems like the first element `dim3[1]` already gets evaluated since the original definition of the variable in the model looks like this:
```
  final parameter Real A3[:,:] = Modelica.Utilities.Streams.readRealMatrix(file3,matrixName1,dim3[1],dim3[2]) "Data of check matrix 3";
```
So it maybe is an oversight when getting the dimension sizes from the output of the function?

**Comment by perost** [comment]
Fixed by #12618.

The issue was that the first time `dim3` got evaluated its binding had to be typed, and this changed its variability from parameter to non-structural parameter since the binding contained calls to external functions. So only the first occurence of `dim3` got evaluated, the rest where skipped since `dim3` had by then been marked as a non-structural parameter.

**Comment by casella** [comment]
Good news: some issues with various versions of MSL were fixed by this: https://libraries.openmodelica.org/branches/history/master/2024-06-19%2013:29:52..2024-06-19%2019:36:09.html. Now 100% of the [latest MSL](https://libraries.openmodelica.org/branches/master/Modelica_trunk/Modelica_trunk.html) 4.1.0 passes the frontend, as well as the subsequent stages up to and including compilation.

![immagine](https://github.com/OpenModelica/OpenModelica/assets/4233148/62e6072d-9cbb-4a93-a74b-03538570dc1a)


Still some issues at runtime, we'll need to handle them. For verification, we are going through a massive update of broken reference files with MAP-LIB, that should help improving the situation.
